pandas/core/dtypes/dtypes.py

```diff
@@ -360,7 +363,8 @@ def __hash__(self) -> int:
             else:
                 return -2
         # We *do* want to include the real self.ordered here
-        return int(self._hash_categories(self.categories, self.ordered))
+        self._hash_cache = int(self._hash_categories(self.categories, self.ordered))
```

Comment:
is _hash_categories ever called with anything other than `(self.categories, self.ordered)`?  if not, could just make it a cache_readonly

Reply:
This is the only place where this function is. To be clear, you propose to do

```
    @cache_readonly
    def __hash__(self):
        ...
```

or would you add this cache to `CategoricalDtype._hash_categories`? `_hash_categories` is a static method and I would assume that breaks since different instances would access the same cache, wouldn't they?

Reply:
the suggestion is that instead of a staticmethod it would become

```
@cache_readonly
def _hash_categories(self):
    categories = self.categories
    ordered = self.ordered 
    [the rest the same as now]
```

Reply:
I tried the `cache_readonly` but ran into two problems

1. Circular imports if imported via `panda.utils`. Could only make this work if I import it directly from the `_libs`
2. I do get failed tests which look like the cache was not attached to an instance but rather the class and/or there is some type confusion. Haven't debugged this further. The failing test is `TestCategoricalDtypeParametrized.test_mixed`

```python
    def test_mixed(self):
        a = CategoricalDtype(["a", "b", 1, 2])
        b = CategoricalDtype(["a", "b", "1", "2"])
>       assert hash(a) != hash(b)
E       AssertionError: assert 7627359294613847363 != 7627359294613847363
E        +  where 7627359294613847363 = hash(CategoricalDtype(categories=['a', 'b', 1, 2], ordered=False))
E        +  and   7627359294613847363 = hash(CategoricalDtype(categories=['a', 'b', '1', '2'], ordered=False))

```

Reply:
> Circular imports if imported via panda.utils. Could only make this work if I import it directly from the _libs

yes do this

Reply:
2. should not be the case, you will have to see why this occurs

Reply:
Was worth looking into this. The cause for 2.) is a name collision with the extension dtype instance caching

https://github.com/pandas-dev/pandas/blob/ce082fb04a0cf597f673d35bc47afd6c2e3c3168/pandas/core/dtypes/dtypes.py#L89

since this and the property cache used the same attribute name `_cache` for their respective caches. Will update the PR shortly

Reply:
xref https://github.com/pandas-dev/pandas/pull/40329